scipy/integrate/_tanhsinh.py

```diff
@@ -373,7 +373,9 @@ def _tanhsinh(f, a, b, *, args=(), log=False, maxfun=None, maxlevel=None,
         a=xp.reshape(a, (-1, 1)), b=xp.reshape(b, (-1, 1)),  # integration limits
         n=minlevel, nit=nit, nfev=nfev, status=status,  # iter/eval counts
         xr0=xr0, fr0=fr0, wr0=wr0, xl0=xl0, fl0=fl0, wl0=wl0, d4=d4,  # err est
-        ainf=ainf, binf=binf, abinf=abinf, a0=xp.reshape(a0, (-1, 1)))  # transforms
+        ainf=ainf, binf=binf, abinf=abinf, a0=xp.reshape(a0, (-1, 1)),  # transforms
+        pc_xjc=None, pc_wj=None, pc_indices=[0], pc_h0=None)  # pair cache
```

Comment:
Please let me know so I can avoid the pattern in the future. I ran into this because it introduced a subtle bug (which I'll fix before the release). 

I understand that global caches can be problematic, but there was a reason for it, so I wonder whether there is an acceptable pattern for maintaining state between calls to a function?

Reply:
If it's easy to explain what the bug is, can you explain? If not, happy to look at the PR when it's ready.

A global cache needs _something_ to make it safe in general. Caches should be protected by a lock or some other way to make them safe for reading/writing in parallel threads. Or made thread-local. Never use a regular global variable for anything unless it's completely immutable.

I don't entirely remember the context since it's 3 weeks ago I looked at it, but this doesn't look like an expensive function call that needs a cache but more like an iterative algorithm that needs the state of the previous iteration. In general, it's cleanest to pass that into the function.

Reply:
> If it's easy to explain what the bug is, can you explain?

Arrays in `work` are subject to being "compressed" - elements corresponding with calculations that have already converged may be eliminated so no further computation is done with them. In some situtations, `pc_xjc`/`pc_wj` were being compressed, but that's not what we want. The simple fix is to protect them by putting them in another container (the `_RichResult` used in `derivative`, for instance). Admittedly, this can also be avoided in the future by separating compressible and incompressible parts of `work` more explicitly, which is something I've planned to do when things slow down a bit.

> this doesn't look like an expensive function call that needs a cache but more like an iterative algorithm that needs the state of the previous iteration. In general, it's cleanest to pass that into the function.

In these cases, the caches were for things like abscissae and weights (e.g. like the outputs of [`roots_legendre`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.roots_legendre.html#scipy.special.roots_legendre)). In this case, the abscissae/weights are mostly independent of the problem being solved... the list of them just grows if the algorithm requires more iterations. The intent was to generate as many as needed the first time the function runs, rather than *every* time, and only ever expand the cache after that. This can be a significant time savings when, say, quadrature is being done within an iterative root-finder (e.g. `ppf` of a distribution that computes the CDF by integrating the PDF). 

The cache *can* change in some conditions. In `tanhsinh`, `h0` is dependent on the dtype, so cache needs to change if different dtype (or array type) is used. In `derivative`, I think they only change if the `step_factor` option is changed. I suppose the user could be working with different dtypes or `step_factor`s in different threads, so it's good to fix, but I'd like to find a way that doesn't reduce the performance in the typical case.

Reply:
Thanks. It sounds like you want something like [`functools.lru_cache`](https://docs.python.org/3/library/functools.html#functools.lru_cache) per dtype then.

For different `step_factor` I'm not sure it makes sense to cache, because the cache size could explore if a user tries different step sizes in a loop (unlike dtypes it's not a fixed set).

Reply:
Just to be clear, it's factors by which the step size is reduced, not the step size itself. This is adaptive, so the user shouldn't need to choose either, but if they do want to try different *initial* step sizes, they can do so without the cache changing at all. (In fact, the code is vectorized w.r.t. initial step, so they can try many in a single call).

When the factor changes, the old cache is replaced. So the cache wouldn't explode if the user loops over step factors.

A problem would (only?) occur if different threads were trying different step factors at the same time. 

I will try `lru_cache`. IIRC that is not thread safe by default, and it requires that arguments are hashable, but it can probably be made to work.

Reply:
> I will try `lru_cache`. IIRC that is not thread safe by default,

Its docs say: _"The cache is threadsafe so that the wrapped function can be used in multiple threads. This means that the underlying data structure will remain coherent during concurrent updates."_

"concurrent updates" (i.e. 2 threads writing at once, possibly the same value, or 1 read and 1 write at once) is normally what thread safety means. Your description of what you want is not that though, and that's because your design sounds like it's not really a cache. A cache has the property that for the same inputs, the outputs are always identical. If that's not the case, you have a bigger issue. If `f(x, y)` returns `42` with the default value of some unknown hidden parameter `z`, and if the user tweaks `z` and the answer becomes `43`, then it sounds like that's inherently not cache-able. `z` must then either become an explicit input parameter so the cache knows about it, or you move the cache so that `z` can no longer influence the result.

> A problem would (only?) occur if different threads were trying different step factors at the same time.

Indeed. It's a bug on any Python version, not only with free-threading. A user can always use the `threading` module.

Reply:
There are a number of explicit tests for concurrency in the code base, named `test_concurrency` or `test_threading` typically. Adopting the test pattern in one of those may help here.